text/0000-combined-restorer-analyzer-phases.md

```diff
@@ -0,0 +1,130 @@
+# Meta
+[meta]: #meta
+- Name: Combined restorer/analyzer and cacher/exporter phases.
+- Start Date: 2019-08-02
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Though not formally part of the spec, the project description and the
+lifecycle reference implementation contain the following four phases:
+
+- `/lifecycle/restorer` (considered an optional platform step)
+- `/lifecycle/analyzer`
+- `/lifecycle/exporter`
+- `/lifecycle/cacher` (considered an optional platform step)
+
+This proposal seeks to combine `restorer` with `analyzer` (to be named
+`restorer`) and `cacher` with `exporter` (to be named `exporter`).
+
+# Motivation
+[motivation]: #motivation
+
+In the current lifecycle reference implementation, the `restorer` phase
+fetches cached layers from a previous cache image. These layers can be
+arbitrarily large and, when using a remote repository, can represent a large
+amount of overhead. The `analyzer` phase runs after `restorer` and it
+may deterine that particular cache layers are no longer valid and remove them
+from the rest of the process - however, we've already paid a large cost to
+fetch them in the first place.
+
+The determination of cache layer validity can frequently be performed
+through the analysis of metadata: comparing metadata from the cache image with
+metadata from the previous application image. However, this comparison can
+only be performed with a combined `restorer` and `analyzer` phase.
+
+A parallel optimization potential exists when combining the `cacher` and
+`exporter` phases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Combined `restorer` phase.
+
+The spec will formalize the (combined) `restorer` phase like this:
+
+```
+/lifecycle/restorer \
+  -group=[(required) group.toml file] \
+  -layers=[(required) layers directory] \
+  -image=[(optional) repository url for existing application image] \
+  -cache=[(optional) local directory, or repository url for cache image]
```

Comment:
I think we should we add a flag for cache type or replace this with two optional flags `cache-volume` and `cache-image`. I worry that at some point a type of caching may be introduced that would prevent us from determining the type of caching from the format of cache destination value.

Reply:
Alternatively, we could provide two `exporter` implementations (one that creates cache image and one for volumes). It might be useful to implement these to demonstrate who the exporter can be extended to support other caching mechanisms.

Reply:
Acknowledged. Unless there are opposing opinions, I will change to mutually exclusive flags `-cache-volume` and `-cache-image`.

Reply:
Not opposed to different exporter binaries, though I'm not sure how one would structure that in the GitHub repo.